storage/integration_test.go

```diff
@@ -3233,7 +3215,100 @@ func TestIntegration_WriterAppend(t *testing.T) {
 				h.mustWrite(w, tc.content)
 
 				// Download content again and validate.
-				// Disabled due to b/408373388; unskip after this is resolved.
+				// Disabled due to b/395944605; unskip after this is resolved.
+				// gotBytes := h.mustRead(obj)
+				// if !bytes.Equal(gotBytes, tc.content) {
+				// 	t.Errorf("content mismatch: got %v bytes, want %v bytes", len(gotBytes), len(tc.content))
+				// }
+
+				// Check object exists and Finalized attribute set as expected.
+				attrs := h.mustObjectAttrs(obj)
+				if tc.finalize && attrs.Finalized.IsZero() {
+					t.Errorf("got unfinalized object, want finalized")
+				}
+				if !tc.finalize && !attrs.Finalized.IsZero() {
+					t.Errorf("got object finalized at %v, want unfinalized", attrs.Finalized)
+				}
+
+			})
+		}
+	})
+}
+
+// Writer test for append takeover of unfinalized object.
+func TestIntegration_WriterAppendTakeover(t *testing.T) {
+	t.Skip("b/402283880")
+	ctx := skipAllButBidi(context.Background(), "ZB test")
+	multiTransportTest(ctx, t, func(t *testing.T, ctx context.Context, _, prefix string, client *Client) {
+		h := testHelper{t}
+		bucketName := prefix + uidSpace.New()
+		bkt := client.Bucket(bucketName)
+
+		h.mustCreateZonalBucket(bkt, testutil.ProjID())
+		defer h.mustDeleteBucket(bkt)
+
+		testCases := []struct {
+			name           string
+			finalize       bool
+			content        []byte
+			chunkSize      int
+			takeoverOffset int64
+		}{
+			{
+				name:           "first message takeover",
+				finalize:       false,
+				content:        randomBytes9MiB,
+				chunkSize:      4 * MiB,
+				takeoverOffset: MiB,
+			},
+			{
+				name:           "first chunk takeover",
+				finalize:       false,
+				content:        randomBytes9MiB,
+				chunkSize:      4 * MiB,
+				takeoverOffset: 3 * MiB,
+			},
+			{
+				name:           "middle chunk takeover",
+				finalize:       false,
+				content:        randomBytes9MiB,
+				chunkSize:      4 * MiB,
+				takeoverOffset: 6 * MiB,
+			},
+			{
+				name:           "final chunk takeover",
```

Comment:
Is it possible to do a zero byte write with FinalizeOnClose= true  to finalize an object? If so, maybe it would be good to test?

Reply:
Also would be good to test an object with takeoverOffset at zero bytes.

Reply:
added first byte and last byte cases as well as takeover just to finalize.